pkg/orchestrator/housekeeping.go

```diff
@@ -191,42 +210,38 @@ func (h *Housekeeping) timeoutExecutions(ctx context.Context, activeExecutions [
 			continue
 		}
 
-		timeoutWithBuffer := execution.Job.Task().Timeouts.GetExecutionTimeout() + h.timeoutBuffer
+		executionTimeout := execution.Job.Task().Timeouts.GetExecutionTimeout()
+		if executionTimeout <= 0 {
+			continue
+		}
+
+		timeoutWithBuffer := executionTimeout + h.timeoutBuffer
 		expirationTime := h.clock.Now().Add(-timeoutWithBuffer)
 		if execution.IsExpired(expirationTime) {
-			// acquire semaphore to limit the number of concurrent housekeeping tasks
-			h.workersSem <- struct{}{}
-			h.waitGroup.Add(1)
 			alreadyEvaluatedJobs[execution.JobID] = struct{}{}
-
-			go func(execution *models.Execution) {
-				defer h.waitGroup.Done()
-				defer func() { <-h.workersSem }() // release semaphore
-				if err := h.handleTimeoutExecutions(ctx, execution); err != nil {
-					log.Ctx(ctx).Err(err).Msgf("failed to handle timeout for execution %s", execution.ID)
-				}
-			}(execution)
+			h.enqueueTimeoutTask(ctx, execution.Job, models.EvalTriggerExecTimeout,
+				fmt.Sprintf("execution %s timed out", execution.ID))
 		}
 	}
 }
 
-// handleTimeoutExecutions handles the timeout of an execution
-// TODO: atomic creation and enqueue of evaluations #3972
-func (h *Housekeeping) handleTimeoutExecutions(ctx context.Context, execution *models.Execution) error {
-	// enqueue evaluation to trigger the scheduler to communicate the cancellation to the compute
-	// node, and schedule a new execution if applicable
-	eval := models.NewEvaluation().
-		WithJobID(execution.JobID).
-		WithTriggeredBy(models.EvalTriggerExecTimeout).
-		WithType(execution.Job.Type).
-		WithComment(fmt.Sprintf("execution %s timed out", execution.ID)).
-		Normalize()
-
-	err := h.jobStore.CreateEvaluation(ctx, *eval)
-	if err != nil {
-		return fmt.Errorf("failed to create evaluation %+v: %w", eval, err)
-	}
-
-	log.Ctx(ctx).Debug().Msgf("enqueued evaluation for timed-out execution %+v", eval)
-	return nil
+func (h *Housekeeping) enqueueTimeoutTask(ctx context.Context, job *models.Job, trigger, comment string) {
+	h.workersSem <- struct{}{}
+	h.waitGroup.Add(1)
+
+	go func() {
```

Comment:
why is it necessary to perform this operation in a go routine?

Reply:
today our datastore is local and may not be that beneficial, but when we use a remote datastore or when we have to replicate datastore operations across multiple requesters, we can benefit from parallelizing unrelated operations or else the housekeeper may not be able to keep up

Reply:
Since we aren't at that point yet this seems unnecessary/premature. But given this PR already merged I suppose it doesn't matter.